profiling/src/collections/string_storage.rs

```diff
@@ -0,0 +1,144 @@
+
+use std::cell::Cell;
+use std::collections::HashMap;
+use std::hash::BuildHasherDefault;
+use std::num::NonZeroU32;
+use std::ptr;
+use std::rc::Rc;
+
+use super::identifiable::StringId;
+use super::string_table::StringTable;
+
+#[derive(PartialEq, Debug)]
+struct ManagedStringData {
+    str: Rc<str>,
+    cached_seq_num: Cell<Option<(*const StringTable, StringId)>>,
+    usage_count: Cell<u32>,
+}
+
+pub struct ManagedStringStorage {
+    next_id: u32,
+    id_to_data: HashMap<u32, ManagedStringData, BuildHasherDefault<rustc_hash::FxHasher>>,
+    str_to_id: HashMap<Rc<str>, u32, BuildHasherDefault<rustc_hash::FxHasher>>,
+    current_gen: u32,
+}
+
+impl ManagedStringStorage {
+    pub fn new() -> Self {
+        let mut storage = ManagedStringStorage {
+            next_id: 0,
+            id_to_data: Default::default(),
+            str_to_id: Default::default(),
+            current_gen: 0,
+        };
+        // Ensure empty string gets id 0 and always has usage > 0 so it's always retained
+        // Safety: On an empty managed string table intern should never fail.
+        storage.intern_new("").expect("Initialization to succeed");
+        storage
+    }
+
+    pub fn advance_gen(&mut self) {
+        self.id_to_data.retain(|_, data| {
+            let retain = data.usage_count.get() > 0;
+            if !retain {
+                self.str_to_id.remove_entry(&data.str);
+            }
+            retain
+        });
+        self.current_gen += 1;
+    }
+
+    pub fn intern(&mut self, item: &str) -> anyhow::Result<u32> {
+        if item.is_empty() {
+            // We don't increase ref-counts on the empty string
+            return Ok(0);
+        }
+
+        let entry = self.str_to_id.get_key_value(item);
+        match entry {
+            Some((_, id)) => {
+                let usage_count = &self
+                    .id_to_data
+                    .get(id)
+                    .ok_or_else(|| {
+                        anyhow::anyhow!("BUG: id_to_data and str_to_id should be in sync")
+                    })?
+                    .usage_count;
+                usage_count.set(usage_count.get() + 1);
+                Ok(*id)
+            }
+            None => self.intern_new(item),
+        }
+    }
+
+    pub fn intern_new(&mut self, item: &str) -> anyhow::Result<u32> {
+        let id = self.next_id;
+        let str: Rc<str> = item.into();
+        let data = ManagedStringData {
+            str: str.clone(),
+            cached_seq_num: Cell::new(None),
+            usage_count: Cell::new(1),
+        };
+        self.next_id = self
+            .next_id
+            .checked_add(1)
+            .ok_or_else(|| anyhow::anyhow!("Ran out of string ids!"))?;
+        let old_value = self.str_to_id.insert(str.clone(), id);
+        debug_assert_eq!(old_value, None);
+        let old_value = self.id_to_data.insert(id, data);
+        debug_assert_eq!(old_value, None);
+        Ok(id)
+    }
```

Comment:
What's the API level difference between these two methods? Also need to document when errors happen, and what maybe to do with them.

Reply:
Ah, nice catch, `intern_new` was fully 100% intended not to be public, yet somehow the `pub` ended up there.

It's not intended to be exposed (too dangerous!) -- I'll remove it.

Reply:
Fixed in 20b946994fa2047d415429d8d65122d1b64a22f5